### PR TITLE
Eva/canvas save button

### PR DESCRIPTION
**Pull request by gevos**

### CODE REVIEW

File: src/components/Checkbox/README.md [hunk 1]
```diff
@@ -3,7 +3,7 @@
 This demonstrates the use of Checkbox.
 
 ```jsx
-import React from  'react';
+import React from 'react';;
```
Comment (edellucien) [inline src/components/Checkbox/README.md]:
```suggestion
import React from 'react';
```

File: src/components/TopicList/README.md [hunk 2]
```diff
@@ -3,7 +3,7 @@
 This demonstrates the use of TopicList.
 
 ```jsx
-import React from  'react';
+import React from 'react';;
```
Comment (edellucien) [inline src/components/TopicList/README.md]:
```suggestion
import React from 'react';
```

File: src/components/CanvasSaveButton/CanvasSaveButton.js [hunk 3]
```diff
@@ -61,53 +98,138 @@ class CanvasSaveButton extends PureComponent {
...
+        if (extraData && extraData.copyright) {
+          destContext.font = '12px Arial';
+          destContext.fillStyle = 'black';
+          destContext.fillText(extraData.copyright, padding, clip.h - padding);
+        }
+
+        // North arrow
+        if (extraData && extraData.northArrow) {
+          const img = new Image();
+          img.src = extraData.northArrow.circled
```
Comment (edellucien) [inline src/components/CanvasSaveButton/CanvasSaveButton.js]: What if we pass a child to NorthArrow, to use it as custom north arrow, such as in this example:
https://github.com/geops/react-spatial/blob/master/src/components/NorthArrow/NorthArrow.test.js#L29
Do you think we should support this scenario here as well?

File: src/components/CanvasSaveButton/CanvasSaveButton.js [hunk 4]
```diff
@@ -61,53 +98,138 @@ class CanvasSaveButton extends PureComponent {
...
+          clip.h,
+        );
+
+        const padding = 5;
+        const arrowSize = 80;
+
+        // Copyright
+        if (extraData && extraData.copyright) {
+          destContext.font = '12px Arial';
+          destContext.fillStyle = 'black';
```
Comment (oterral) [inline src/components/CanvasSaveButton/CanvasSaveButton.js]: These properties should also come from the extraData object.

File: src/components/CanvasSaveButton/README.md [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+    const extent = [-16681616.919511989, 655523.1517989757, -5312679.080488012, 8482674.848201025];
+
+    return (
+      <div className="tm-canvas-save-button-example">
+        <BasicMap
+          map={this.map}
+          center={this.center}
+          zoom={3}
+        />
+        <ShareMenu url={window.location.href}>
```
Comment (oterral) [inline src/components/CanvasSaveButton/README.md]: I don't think you need the share menu in this example. CanvasSaveButton can work standalone.

File: src/components/CanvasSaveButton/README.md [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+          <CanvasSaveButton
+            title="Karte als Bild speichern."
+            className='tm-canvas-save-button tm-share-menu-icon'
+            map={this.map}
+            extent={extent}
+            extraData={{
+              copyright: '© Example copyright',
```
Comment (oterral) [inline src/components/CanvasSaveButton/README.md]: it should be possible to get the copyright from the layerService like this:
```
copyright: layerService.getCopyrights()
```
Does it make sense?

Comment (oterral) [reply src/components/CanvasSaveButton/README.md]: Can you try to set the copyright dynamically ? if a layers change is visibility the copyright must be updated.

File: src/components/CanvasSaveButton/README.md [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+              northArrow: {
+                rotation: 25,
```
Comment (oterral) [inline src/components/CanvasSaveButton/README.md]: Same here the `NorthArrow` component can have a `rotateOffset`, see `NorthArrow` component.
Can you try to set the good rotation of the north arrow  dynamically ?
if the map is rotated the rotation must be updated too.

File: src/components/CanvasSaveButton/README.md (hunk 6 above)
Comment (gevos) [reply src/components/CanvasSaveButton/README.md]: I had implemented this, see https://github.com/geops/react-spatial/pull/62/commits/12f9038758ad4a89258bf096b63ced95731ed8b6#diff-84392269f1dcee3ac19b57cd31741243L145. However, after I've talked to @mhaertwig, yesterday, I decided to simplify the properties. So the canvas save button is only responsible for writing on the canvas. Extracting the current copyright and retrieving the map rotation is not handled anymore by the canvas save button.

File: src/components/CanvasSaveButton/README.md (hunk 7 above)
Comment (gevos) [reply src/components/CanvasSaveButton/README.md]: See above

File: src/components/CanvasSaveButton/README.md (hunk 6 above)
Comment (mhaertwig) [reply src/components/CanvasSaveButton/README.md]: @oterral I think we should keep the component simple. It's good enough to define the copyright manually here. 

However, if we want to reuse the functionality to display copyrights of active layers, we could add a `getActiveLayerCopyrights()` function to the `LayerService` that returns a list of visible layer copyrights.

File: src/components/CanvasSaveButton/CanvasSaveButton.js (hunk 4 above)
Comment (gevos) [reply src/components/CanvasSaveButton/CanvasSaveButton.js]: Done.

File: src/components/CanvasSaveButton/README.md (hunk 5 above)
Comment (gevos) [reply src/components/CanvasSaveButton/README.md]: Done.

File: src/components/CanvasSaveButton/README.md (hunk 6 above)
Comment (oterral) [reply src/components/CanvasSaveButton/README.md]: That's what I propose in my first comment. I just want to be sure it works , that the component is rerendered when a copyright change. 
Or even better, maybe just manage the case if this copyright property is a function.
```
extraData={{
    copyright: () => {
       return  layerService.getCopyrigths();
   }
}]

File: src/components/CanvasSaveButton/README.md (hunk 7 above)
Comment (oterral) [reply src/components/CanvasSaveButton/README.md]: Same answer so :) . We should be able to set a function in the `rotation` property. To calculate the good rotation dynamically.

File: src/components/CanvasSaveButton/CanvasSaveButton.js (hunk 3 above)
Comment (oterral) [reply src/components/CanvasSaveButton/CanvasSaveButton.js]: he's right,`northArrow` to display  should come from the properties of the object as well.

Comment (gevos) [reply src/components/CanvasSaveButton/CanvasSaveButton.js]: > Changes I propose :
>     * [...]
>     * Add properties src, width, height to northArrow property to be able to specify a custom northArrow.

Done with src, width, and height

File: src/components/CanvasSaveButton/README.md (hunk 7 above)
Comment (gevos) [reply src/components/CanvasSaveButton/README.md]: > Changes I propose :
> 
>     * [...]
> 
>     * Add the possibility to specify a function or a number to the rotation property.

Implemented. Now, you can pass a rotation function or a number

File: src/components/CanvasSaveButton/README.md (hunk 6 above)
Comment (gevos) [reply src/components/CanvasSaveButton/README.md]: > Changes I propose :
> 
>     * Add the possibility to specify a function or a number to the copyright property.
Implemented. Now you can pass a copyright function or a text.

I had re-reverted the static getCopyrights() function from the Copyright component. It is used in the CanvasSaveButtonExample. @oterral  suggested in the example above to define a getCopyrights() function in the LayerService. @edellucien @mhaertwig, should I leave it in the Copyright component or should I move it to the LayerService?

**Comment by gevos** [comment]
> Just make an example a bit more dynamic where we can change the visibility of a layer,

Done. By using the getCopyrights() function, only the currently visible copyrights are returned.

File: src/components/CanvasSaveButton/README.md (hunk 6 above)
Comment (edellucien) [reply src/components/CanvasSaveButton/README.md]: Das Endergebnis ist die gleiche, aber ich denke auch, es wäre sauberer, es in layerService zu finden. Stimmst du auch dazu @mhaertwig ?

**Pull request merged**